Design Guide: Designing ZTNA access policies for Cloudflare Access (#18437)
cloudflare/cloudflare-docs: src/content/docs/reference-architecture/design-guides/designing-ztna-access-policies.mdx

- **Is the request being made by another process or application?**
  It is not always a real human on a device attempting to access an application. This makes it useful to leverage Cloudflare Access to manage communication to APIs by other software. The request may contain service tokens, mutual TLS certificates, and SSH certificates, which enables logins for automated processes and machine-to-machine communication. Using service auth options within Cloudflare also centralizes the storage and lifecycle management of these tokens and certificates.

- **What does your third-party tool say about your device?**
  Many organizations use other specialized tools for endpoint security, such as Crowdstrike, SentinelOne, or Microsoft Intune, to provide telemetry regarding the security posture of the device making the application request. Rather than require the user to navigate multiple UIs, you can integrate these tools into Cloudflare One via their API, and apply their insights into device posture attributes that can be enforced during an application login.

:::note[]

> Some third party device posture integrations can be used even when the user device doesn't have our agent installed. Instead, the third party integration matches the user based on email and provides information directly to Cloudflare.

Reviewer: ```suggestion
Some third-party device posture integrations can be used even when the user device does not have our agent installed. Instead, the third party integration matches the user based on email and provides information directly to Cloudflare.
```

:::

#### Additional settings

Below are a few additional application settings to consider that help improve security.

##### Isolate application

Sometimes you want to manage access to a self-hosted application for less trusted, third-party users such as contractors or partners. You might want to allow them to read content in an application, but limit their ability to download files, copy and paste data, and print the page. Cloudflare Access allows you to render the application in a remote [browser](/cloudflare-one/policies/access/isolate-application/) (using [remote browser isolation, or RBI](/cloudflare-one/policies/browser-isolation/)) so that the application is rendered using a headless browser on our network versus sending all the content down to the user's browser. This allows Cloudflare to then enforce a range of controls over how the user can interact with the content.

The setting is at the policy level, so one policy can allow trusted users (such as employees) to access applications normally, while another policy with browser isolation enabled can apply the RBI service for contractors.

> This setting forces traffic to an isolated browser before being delivered to the end user, which means all traffic is then inspected and managed by Cloudflare Gateway. To limit what the user can do, you need to create an accompanying policy in the gateway, which also identifies the same users and then enforces the [controls](/cloudflare-one/policies/browser-isolation/isolation-policies/#policy-settings) you wish to limit access. Note that it's important to write the Gateway policy such that it only enforces RBI for the same group of users accessing the application that the Cloudflare Access policy applies to. Otherwise, the policy will default to enforce browser isolation for all users.

Reviewer: ```suggestion
This setting forces traffic to an isolated browser before being delivered to the end user, which means all traffic is then inspected and managed by Cloudflare Gateway. To limit what the user can do, you need to create an accompanying policy in the gateway, which also identifies the same users and then enforces the [controls](/cloudflare-one/policies/browser-isolation/isolation-policies/#policy-settings) you wish to limit access. Note that it is important to write the Gateway policy such that it only enforces RBI for the same group of users accessing the application that the Cloudflare Access policy applies to. Otherwise, the policy will default to enforce browser isolation for all users.
```

![Figure 6 - An access group that matches to IT administrators on secure systems.](~/assets/images/reference-architecture/designing-ztna-access-policies-for-cloudflare-access/figure6.svg "Figure 6 - An access group that matches to IT administrators on secure systems.")

## Use cases

Now that the basic infrastructure to secure access to an application and the policy systems have been covered, let's dive into some common use cases.

### Only allow company wiki access to users on trusted devices

Many companies host some sort of internal content system where confidential company information resides. Wiki's are a common type of application that allows employees to collaborate easily with anyone. But because this information is confidential, it is important to both validate the user authentication with strong credentials, and also ensure that their access is via a secure device, via a secure connection.

> However, sometimes company users use non-company devices and need to access the wiki. You may wish to set up a policy that allows this, but limits the user's actions; for instance, prevent them from editing the data, or from copying and pasting it to their unmanaged device. This use case explains how to set up a Cloudflare Access application to define secure access for employees, giving them fully functional access when they are on a secured device over a secured connection, but still allow them some limited access from a non secure device.

Reviewer: ```suggestion
However, sometimes company users use non-company devices and need to access the wiki. You may wish to set up a policy that allows this, but limits the user's actions. For instance, prevent them from editing the data, or from copying and pasting it to their unmanaged device. This use case explains how to set up a Cloudflare Access application to define secure access for employees, giving them fully functional access when they are on a secured device over a secured connection, but still allow them some limited access from a non secure device.
```

First, create an Access application with the following parameters:

| Name            | Company Wiki                                                   |
| :-------------- | :------------------------------------------------------------- |
| Type            | Self-hosted                                                    |
| Public Hostname | wiki.mycustomerexample.com                                     |
| Authentication  | Company Microsoft Entra IdP                                    |
| Policies        | Employees on trusted devices Employees using untrusted devices |

> Before we examine how the two policies are defined, let's look at an Access Group created to identify an employee and approved devices running the latest operating system version.

Reviewer: ```suggestion
Before we examine how the two policies are defined, observe an example where an Access Group was created to identify an employee and approved devices were running the latest operating system version.
```

#### Employees on trusted devices

Now we define the first policy in the application. First, select the Access Group that has already been defined. Then, define the following rules to determine how users authenticate and how they connect to the application.

| Policy name             | Employees on trusted devices          |
| :---------------------- | :------------------------------------ |
| Action                  | Allow                                 |
| Access groups           | Include \- Secure employees           |
| **Rules**               |                                       |
| Require                 |                                       |
| Authentication Method   | MFA \- Multiple Factor Authentication |
| Gateway                 | On                                    |
| **Additional settings** |                                       |
| Isolate Application     | No                                    |

> This policy ensures that users can gain full access to your company wiki only if they have passed the following requirements: they are full-time employees on devices with the latest operating system; have authenticated using MFA; and are accessing the application via a device that has the Cloudflare device agent running.

Reviewer: ```suggestion
This policy ensures that users can gain full access to your company wiki only if they have passed the following requirements: 

- They are full-time employees on devices with the latest operating system.
- Users have authenticated using MFA.
- Users are accessing the application via a device that has the Cloudflare device agent running.
```

Essentially, the employee on an insecure device is permitted to view the wiki in a "read-only" mode, but is restricted from further interactions like uploading/downloading or copying/pasting confidential information.

This policy approach accomplishes several objectives:

1. It enforces the use of trusted devices for full access to the wiki, aligning with your Zero Trust security goals.
2. It provides a fallback option for employees using personal devices, allowing them to access the wiki in a limited, secure manner through browser isolation.
3. It incentivizes employees to use their company devices and/or keep WARP enabled, which is a net positive for an organization's security posture.
4. It demonstrates the power and flexibility of more granular security controls achieved by combining Cloudflare Access policies with Cloudflare Gateway HTTP policies.

This approach both secures your wiki and establishes a model for protecting other applications — allowing your organization to maintain strong cyber hygiene while adapting to the realities of hybrid work scenarios.

### Secure access to Salesforce

> The second use case implements a secure access strategy that also requires the use of the device client; however, the implementation is slightly more involved than the previous wiki example.

Reviewer: ```suggestion
The second use case implements a secure access strategy that also requires the use of the device client. However, the implementation is slightly more involved than the previous wiki example.
```

> Before addressing the specifics, let's consider the benefits of securing access to SaaS apps through Cloudflare. After all, Salesforce and other major SaaS providers already offer robust security features, including their own access controls, MFA, and audit logs. So why do some organizations still choose to route their SaaS traffic through Cloudflare?

Reviewer: ```suggestion
Before addressing the specifics, you will learn about the benefits of securing access to SaaS apps through Cloudflare. After all, Salesforce and other major SaaS providers already offer robust security features, including their own access controls, MFA, and audit logs. So why do some organizations still choose to route their SaaS traffic through Cloudflare?
```

> The key benefit here is centralizing security policy enforcement across your entire IT ecosystem. By routing Salesforce access through Cloudflare, you're not just securing Salesforce – you're integrating it into a broader Zero Trust strategy that includes a single point of visibility for all user activity, and reduces the complexity of managing multiple security systems. It also allows you to implement the enforcement of many different IdPs for access to a single SaaS application.

Reviewer: ```suggestion
The key benefit here is centralizing security policy enforcement across your entire IT ecosystem. By routing Salesforce access through Cloudflare, you are not just securing Salesforce – you are integrating it into a broader Zero Trust strategy that includes a single point of visibility for all user activity, and reduces the complexity of managing multiple security systems. It also allows you to implement the enforcement of many different IdPs for access to a single SaaS application.
```

> In the context of this use case, it's important to protect Salesforce — which contains sensitive customer data — against misuse, and to secure access only to authorized users. We're going to design a secure access policy that can cover both of these objectives.

Reviewer: ```suggestion
In the context of this use case, it is important to protect Salesforce — which contains sensitive customer data — against misuse, and to secure access only to authorized users. We are going to design a secure access policy that can cover both of these objectives.
```

The first step is to configure an [egress IP policy under Cloudflare Gateway](/cloudflare-one/policies/gateway/egress-policies/). This allows you to purchase and assign specific IPs to your users that have their traffic filtered via Gateway. Then in Salesforce, you can enforce that access is only permitted for traffic with a source IP that matches the one in your egress policy. This combination ensures that the only way to get access to Salesforce is via Cloudflare.

| Egress Policy                       |                  |
| :---------------------------------- | :--------------- |
| **Identity**                        |                  |
| User Group Names                    | All Employees    |
| **Select Egress IP**                |                  |
| Use dedicated Cloudflare Egress IPs | \[203.0.113.88\] |

> This is important not only for securing access to Salesforce, but also for adequately protecting its contents while in use. Now let's look at the access policy, similar to the one we just created for the wiki. But now the policy is limiting access to members of the Sales or Executives groups. We are also using our Crowdstrike integration to ensure that users are on company managed devices.

Reviewer: ```suggestion
This is important not only for securing access to Salesforce, but also for adequately protecting its contents while in use. The next step is to examine the access policy. The policy is similar to the one we just created for the wiki. However, the policy is limiting access to members of the Sales or Executives groups. We are also using our Crowdstrike integration to ensure that users are on company managed devices.
```

| Crowdstrike Service to Service | Overall Score above 80                |

The second policy now applies to all employees but we are going to apply a few more steps before access is granted.

| Policy name                    | Employees on trusted devices       |
| :----------------------------- | :--------------------------------- |
| Action                         | Allow                              |
| **Include**                    |                                    |
| Member of group                | All Employees                      |
| **Require**                    |                                    |
| Authentication method          | MFA \- multi-factor authentication |
| Gateway                        | On                                 |
| Crowdstrike Service to Service | Overall Score above 80             |
| **Additional Settings**        |                                    |
| Purpose justification          | On                                 |
| Temporary authentication       | On                                 |
| Email addresses of approvers   | salesforce-admin@company.com       |

> We’re going to add in temporary authentication to this second policy. That means if Cloudflare determines that the incoming request is from someone outside of the Sales or Executives department, an administrator will need to explicitly grant them temporary access. In context, this policy could be used to secure access to Salesforce for employees outside the Sales department, as the customer information could be sensitive and confidential.

Reviewer: ```suggestion
We are going to add in temporary authentication to this second policy. That means if Cloudflare determines that the incoming request is from someone outside of the Sales or Executives department, an administrator will need to explicitly grant them temporary access. In context, this policy could be used to secure access to Salesforce for employees outside the Sales department, as the customer information could be sensitive and confidential.
```

| :----------------------------------- | :------------------------------------ |
| Action                               | Allow                                 |
| **Include**                          |                                       |
| Assign a group                       | IT Admins                             |
| **Require**                          |                                       |
| Authentication method                | MFA \- multi-factor authentication    |
| Gateway                              | On                                    |
| Device Posture \- Serial Number List | Company Managed Device Serial Numbers |
| OS Version                           | Latest version of Windows             |
| Domain Joined                        | Joined to corporate AD domain         |
| **Exclude**                          |                                       |
| Authentication method                | SMS                                   |
| **Additional Settings**              |                                       |
| Purpose justification                | On                                    |

> Here, we're introducing a high number of security posture checks, starting with MFA. We have two expressions regarding MFA: the first one requires that users authenticate with a MFA method; the second 'excludes' expression pointing out that SMS is not considered a valid authentication method. We do this because SMS is one of the easier methods for attackers to exploit and subvert, and therefore [considered less secure](https://sec.okta.com/articles/2020/05/sms-two-factor-authentication-worse-just-good-password) than other MFA methods. As a result, we're only allowing access when the user provides stronger credentials such as a hard key or an OTP from an authenticator app. Enforcing these stricter MFA requirements reduces the risk of credential-based attacks, and makes it much more challenging for potential attackers to gain unauthorized access to this critical database—even if they have obtained the user's password.

Reviewer: ```suggestion
Here, we are introducing a high number of security posture checks, starting with MFA. We have two expressions regarding MFA: the first one requires that users authenticate with a MFA method. The second 'excludes' expression pointing out that SMS is not considered a valid authentication method. We do this because SMS is one of the easier methods for attackers to exploit and subvert, and therefore [considered less secure](https://sec.okta.com/articles/2020/05/sms-two-factor-authentication-worse-just-good-password) than other MFA methods. As a result, we are only allowing access when the user provides stronger credentials such as a hard key or an OTP from an authenticator app. Enforcing these stricter MFA requirements reduces the risk of credential-based attacks, and makes it much more challenging for potential attackers to gain unauthorized access to this critical database—even if they have obtained the user's password.
```

> Other posture elements here include requiring the latest OS; that the user's device is joined to a Microsoft Active Directory domain; and that their device is explicitly a company-managed device (shown by referencing a list of managed device serial numbers). These combined posture checks ensure that only up-to-date, company-controlled devices within your managed environment can access the database, further reducing the attack surface and the risk of access from potentially compromised or uncontrolled endpoints.

Reviewer: ```suggestion
Other posture elements here include:

- Requiring the latest OS.
- The user's device is joined to a Microsoft Active Directory domain.
- The user's device is explicitly a company-managed device (shown by referencing a list of managed device serial numbers). 

These combined posture checks ensure that only up-to-date, company-controlled devices within your managed environment can access the database, further reducing the attack surface and the risk of access from potentially compromised or uncontrolled endpoints.
```

> Under additional settings, we're also requiring that users enter a purpose justification for accessing the database. This allows your security teams to analyze access patterns and identify potentially suspicious behavior. This set of security controls also ensures that access to your critical database is tightly regulated, logged, and justified — significantly reducing the risk of unauthorized access or misuse.

Reviewer: ```suggestion
Under additional settings, we are also requiring that users enter a purpose justification for accessing the database. This allows your security teams to analyze access patterns and identify potentially suspicious behavior. This set of security controls also ensures that access to your critical database is tightly regulated, logged, and justified — significantly reducing the risk of unauthorized access or misuse.
```

This level of protection and visibility would be significantly more complex and resource-intensive to achieve with disparate, standalone security solutions. Centralizing security policy enforcement via Cloudflare allows you to simplify how you implement fine-grained access to critical internal resources.

### Secure RDP access

This final use case centers on securing remote access to devices via RDP in two ways — self-hosted or private IP. Both options offer unique benefits, but ultimately it comes down to your priorities: is it more important to simplify access, or to tightly monitor activity?

> We'll start with the self-hosted option — proxying port 3389 over a tunnel, mapping it to a hostname.

Reviewer: ```suggestion
We will start with the self-hosted option — proxying port 3389 over a tunnel, mapping it to a hostname.
```

| Application Configuration |                                     |
| :------------------------ | :---------------------------------- |
| Application Name          | RDP service on database server      |
| Hostname                  | rdp.databaseserver.company.internal |

> And now let's define the policy.

Reviewer: ```suggestion
Define the policy:
```

| Policy name                          | Admin Access                          |
| :----------------------------------- | :------------------------------------ |
| Action                               | Allow                                 |
| **Include**                          |                                       |
| Member of Group                      | IT Admins                             |
| **Require**                          |                                       |
| Authentication method                | MFA \- multi-factor authentication    |
| Gateway                              | On                                    |
| WARP                                 | On                                    |
| Device Posture \- Serial Number List | Company Managed Device Serial Numbers |
| External Evaluation                  | \[Time Evaluator URL\]                |

> Inside the policy, we've made this application available to our new access group for IT Admins. Under "Require," we're enforcing the use of Cloudflare WARP specifically (as opposed to only Cloudflare Gateway); the user must be on a company-managed device, with an active device client that's authenticated to the company's instance of Cloudflare, MFA must be used during login, and there is an additional option below for external evaluation.

Reviewer: ```suggestion
Inside the policy, we have made this application available to our new access group for IT Admins. Under "Require," we are enforcing the use of Cloudflare WARP specifically (as opposed to only Cloudflare Gateway). The user must be on a company-managed device, with an active device client that is authenticated to the company's instance of Cloudflare, MFA must be used during login, and there is an additional option below for external evaluation.
```

> [External evaluation](/cloudflare-one/policies/access/external-evaluation/) means we have an API endpoint containing some sort of [access logic](https://github.com/cloudflare/workers-access-external-auth-example) — in this case, time of day access. We're making an API call to this endpoint, and defining the key that Cloudflare is using to verify that the response came from the API. This is useful for several reasons:

Reviewer: ```suggestion
[External evaluation](/cloudflare-one/policies/access/external-evaluation/) means we have an API endpoint containing some sort of [access logic](https://github.com/cloudflare/workers-access-external-auth-example) — in this case, time of day access. We are making an API call to this endpoint, and defining the key that Cloudflare is using to verify that the response came from the API. This is useful for several reasons:
```

> External evaluation allows users to create bespoke security posture checks based on criteria that may not be covered by the default set of posture checks. For this example, we'd be using a service built on [Cloudflare Workers](https://workers.cloudflare.com/).

Reviewer: ```suggestion
External evaluation allows users to create bespoke security posture checks based on criteria that may not be covered by the default set of posture checks. For this example, we will be using a service built on [Cloudflare Workers](https://workers.cloudflare.com/).
```

- Restricting access to the terminal outside of business hours implements a form of time-based access control. This adds an extra layer of security by limiting the window of opportunity for potential attackers.

> Now, let’s look at how you’d secure RDP access as a private IP application.

Reviewer: ```suggestion
Now, you will learn how to secure RDP access as a private IP application:
```

| Application Configuration |                 |
| :------------------------ | :-------------- |
| Application Name          | RDP             |
| Destination IP            | 169.254.255.254 |

> As mentioned before, private IP applications work because Cloudflare proxies the IP range across its network. The nature of this application necessitates the use of the device client, as unless the user is connected to Cloudflare (and more specifically, unless they can take advantage of the WARP-to-Tunnel connectivity), they won’t be able to reach non-local RFC 1918 addresses.

Reviewer: ```suggestion
As mentioned before, private IP applications work because Cloudflare proxies the IP range across its network. The nature of this application necessitates the use of the device client, as unless the user is connected to Cloudflare (and more specifically, unless they can take advantage of the WARP-to-Tunnel connectivity), they will not be able to reach non-local RFC 1918 addresses.
```

| Traffic                                  |                                                                 |
| :--------------------------------------- | :-------------------------------------------------------------- |
| Destination IP                           | 169.254.255.254                                                 |
| Destination Port                         | 3389                                                            |
| **Identity**                             |                                                                 |
| User Group Names                         | Server Admins                                                   |
| **Device Posture**                       |                                                                 |
| Passed Device Posture Checks             | WARP Check (Mac OS) (File) Latest Version of macOS (OS version) |
| **Action**                               | Allow                                                           |
| **Enforce WARP client session duration** | 60m0s                                                           |

> Defining the application here is simple, as Cloudflare automatically fills in the IP range, and you need to limit the detected protocol to RDP. However, the rules for private IP applications are slightly different. You'll notice they appear as network policies under the Cloudflare Gateway menu, despite managing them in Access. Certain options, such as checking for MFA and external evaluation, do not appear here. However, these attributes can be verified when the user activates their device client and authenticates to their organization.

Reviewer: ```suggestion
Defining the application here is simple, as Cloudflare automatically fills in the IP range, and you need to limit the detected protocol to RDP. However, the rules for private IP applications are slightly different. You will notice they appear as network policies under the Cloudflare Gateway menu, despite managing them in Access. Certain options, such as checking for MFA and external evaluation, do not appear here. However, these attributes can be verified when the user activates their device client and authenticates to their organization.
```

> One option available here is enforcing the device agent client session duration. This means that after a certain amount of time, the user will be forced to reauthenticate. This feature allows you to take a Zero Trust approach to protecting private IP applications as well; it ensures that even if a user's credentials are compromised or their device is left unattended, the potential window for unauthorized access is limited. By regularly requiring reauthentication, we're continuously verifying the user's identity and authorization status, aligning with the core Zero Trust principle of "never trust, always verify."

Reviewer: ```suggestion
One option available here is enforcing the device agent client session duration. This means that after a certain amount of time, the user will be forced to reauthenticate. This feature allows you to take a Zero Trust approach to protecting private IP applications as well. It ensures that even if a user's credentials are compromised or their device is left unattended, the potential window for unauthorized access is limited. By regularly requiring reauthentication, we are continuously verifying the user's identity and authorization status, aligning with the core Zero Trust principle of "never trust, always verify."
```

By combining granular access controls with detailed activity logging, Cloudflare provides a comprehensive security solution for protecting and monitoring access to critical resources in a Zero Trust methodology.

## Summary

Successful ZTNA implementation is about more than just technical configuration — it requires careful consideration of your organization's specific needs, user workflows, and security requirements. Cloudflare's flexibility allows you to start with basic secure access policies, then evolve them as your organization's needs change and security requirements mature. By following the principles and practices outlined in this guide, you can create a robust security posture that protects you as precisely and transparently as possible.

> If you're interested in learning more about ZTNA, SASE, or other aspects of the Cloudflare One platform, please visit our [reference architecture library](/reference-architecture/) or our [developer docs](/) to get started.

Reviewer: ```suggestion
If you are interested in learning more about ZTNA, SASE, or other aspects of the Cloudflare One platform, please visit our [reference architecture library](/reference-architecture/) or our [developer docs](/) to get started.
```
